keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,492 @@
+kind: CustomResourceDefinition
+...
+  schema:
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          x-kubernetes-validator: 
+            - rule: "minReplicas <= maxReplicas"
+              message: "minReplicas cannot be larger than maxReplicas"
+          type: object
+          properties:
+            minReplicas:
+              type: integer
+            maxReplicas:
+              type: integer
+```
+
+- Each validator may have multiple validation rules.
+
+- Each validation rule has an optional 'message' field for the error message that
+will be surfaced when the validation rule evaluates to false.
+
+- The validator will be scoped to the location of the `x-kubernetes-validator`
+extension in the schema. In the above example, the validator is scoped to the
+'spec' field.
+
+- For OpenAPIv3 object types, the expression will have direct access to all the
+fields of the object the validator is scoped to.
+  
+- For OpenAPIv3 scalar types (integer, string & boolean), the expression will have
+access to the scalar data element the validator is scoped to.
+
+- For OpenAPIv3 list and map types, the expression will have access to the data
+element of the list or map.
+  
+- For immutability use case, validator will have access to the existing version of the object. 
+
+<<[UNRESOLVED @jpbetz @deads2k @lavalamp]>>
+When performing an update, should the live state of a field also be readable by expressions? This could be used,
+for example, to enforce immutability.
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED @cici37]>>
+Should the message also be an expression to allow for some basic variable
+substitution?
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED @cici37]>>
+Should a 'type' field also be required? If it is not 'cel', the validator
+could be skipped to future proof against addition of other ways to validate in
+the future, or to allow 3rd party validators to have a way to inline their
+validation rules in CRDs.
+<<[/UNRESOLVED]>>
+
+#### Field paths and field patterns
+
+A field path is a patch to a single node in the data tree. I.e. it specifies the
+exact indices of the list items and the keys of map entries it traverses.
+
+A field *pattern* is a path to all nodes in the data tree that match the pattern. I.e.
+it may wildcard list item and map keys.
+
+
+#### Expression lifecycle
+
+When CRDs are written to the kube-apiserver, all expressions will be [parsed and
+typechecked](https://github.com/google/cel-go#parse-and-check) and the resulting
+program will be cached for later evaluation (CEL evaluation is thread-safe and
+side-effect free). Any parsing or type checking errors will cause the CRD write
+to fail with a descriptive error.
+
+#### Function library
+
+The function library available to expressions can be augmented using [extension
+functions](https://github.com/google/cel-spec/blob/master/doc/langdef.md#extension-functions).
+
+List of functions to include for the initial release:
+- Equality and Ordering
```

Comment:
equality doesn't appear to reflect our semantic equality for lists that are keyed by a string

Reply:
Good catch, I've added note here so we don't forget to handle semantic equality.  (I'll also review CEL and make sure it doesn't do anything special with equality that we're going to have to override or work around)

Reply:
Sorry I didn't quite understand this point. I thought the definition of equality(dictionary/list) in CEL is semantic equality from [CEL documentation](https://github.com/google/cel-spec/blob/master/doc/langdef.md#equality-and-ordering). Would you mind to give an example on this? 🙏

Reply:
The rule defined for CEL for list equality ("a[0] == b[0] && a[1] == b[1] && ... && a[N-1] == b[N-1]") would consider a Kubernetes associative lists to be not equal even when Kubernetes semantic equality considers them to be equal because Kubernetes associative list equality uses map equality semantics (where one of the fields is known to be the key) and the order of the list is irrelevant.

Reply:
Thanks for the clarification. I will add it into the kep